cmd/query/app/apiv3/otlp_translator.go

```diff
@@ -0,0 +1,364 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apiv3
+
+import (
+	"encoding/binary"
+	"fmt"
+	"hash/fnv"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/collector/translator/conventions"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+
+	"github.com/jaegertracing/jaeger/model"
+	commonv1 "github.com/jaegertracing/jaeger/proto-gen/otel/common/v1"
+	resourcev1 "github.com/jaegertracing/jaeger/proto-gen/otel/resource/v1"
+	v1 "github.com/jaegertracing/jaeger/proto-gen/otel/trace/v1"
+)
+
+// OpenTelemetry collector implements translator from Jaeger model to pdata (wrapper around OTLP).
+// However, it cannot be used because the imported OTLP in the translator is in the collector's private package.
+func jaegerSpansToOTLP(spans []*model.Span) []*v1.ResourceSpans {
+	spansByLibrary := make(map[resource]map[instrumentationLibrary]*v1.InstrumentationLibrarySpans)
+	for _, s := range spans {
+		otlpSpan, res, library := jSpanToOTLP(s)
+		resourceSpans, ok := spansByLibrary[res]
+		if !ok {
+			resourceSpans = map[instrumentationLibrary]*v1.InstrumentationLibrarySpans{}
+			resourceSpans[library] = &v1.InstrumentationLibrarySpans{
+				InstrumentationLibrary: &commonv1.InstrumentationLibrary{
+					Name:    library.name,
+					Version: library.version,
+				},
+			}
+			spansByLibrary[res] = resourceSpans
+		}
+		resourceSpans[library].Spans = append(resourceSpans[library].GetSpans(), otlpSpan)
+	}
+
+	var rss []*v1.ResourceSpans
+	for res, libMap := range spansByLibrary {
+		rs := &v1.ResourceSpans{
+			Resource: res.resource,
+		}
+		for _, v := range libMap {
+			rs.InstrumentationLibrarySpans = append(rs.InstrumentationLibrarySpans, v)
+		}
+		rss = append(rss, rs)
+	}
+	return rss
+}
+
+type instrumentationLibrary struct {
+	name, version string
+}
+
+// helper type used as a map key
+type resource struct {
+	serviceName string
+	// concatenated string tags to make sure services with the same name but different versions
```

Comment:
why does the comment specifically mention versions? The code seems to suggest that any process tag could be a differentiator, along with versions, such as ip, host.name, etc.?

Reply:
I will edit the comment. The version was just an example.